### PR TITLE
First post needs moderation

### PR DESCRIPTION
**Pull request by dom96**

### CODE REVIEW

File: editdb.nim [hunk 1]
```diff
@@ -1,11 +1,13 @@
...
-var db = Open(connection="nimforum.db", user="postgres", password="", 
+var db = open(connection="nimforum.db", user="postgres", password="",
               database="nimforum")
 
-db.exec(sql"""ALTER TABLE person add column
-  lastOnline timestamp
-""", [])
+db.exec(sql("update person set status = ?"), $User)
+db.exec(sql("update person set status = ? where ban <> ''"), $Troll)
+db.exec(sql("update person set status = ? where ban like '%spam%'"), $Spammer)
```
Comment (dom96) [inline editdb.nim]: Why do we have a separate rank for spammers and trolls? What's the point?

File: forms.tmpl [hunk 2]
```diff
@@ -1,12 +1,19 @@
...
 #
-#template `%`(idx: expr): expr {.immediate.} =
+#template `%`(idx: untyped): untyped =
 #  row[idx]
 #end template
 #
 #
 #proc genThreadsList(c: var TForumData, count: var int): string =
-#  const query = sql"select id, name, views, modified from thread order by modified desc limit ?, ?"
+#  const queryAdmin = sql"""select id, name, views, modified from thread
```
Comment (dom96) [inline forms.tmpl]: This should be called ``queryModAdmin`` ;)

File: forms.tmpl [hunk 3]
```diff
@@ -1,12 +1,19 @@
...
+#                    where id in (select thread from post where author in
+#                             (select id from person where status not in ('Spammer') or id = ?))
```
Comment (dom96) [inline forms.tmpl]: a 3 deep subquery? Won't this run like a snail?

File: forms.tmpl [hunk 4]
```diff
@@ -1,12 +1,19 @@
...
+#                           order by modified desc limit ?, ?"""
+#  const query = sql"""select id, name, views, modified from thread
+#                    where id in (select thread from post where author in
+#                             (select id from person where status not in ('Moderated', 'Spammer') or id = ?))
```
Comment (dom96) [inline forms.tmpl]: There is a lot of duplication between these two queries, could you reduce that?

File: forms.tmpl [hunk 5]
```diff
@@ -34,7 +41,8 @@
   </div>
 </div>
 <div id="talk-threads">
-#  for row in rows(db, query, $((c.pageNum-1) * ThreadsPerPage), $ThreadsPerPage):
+#  for row in rows(db, if c.rank >= Moderator: queryAdmin else: query,
+#                  c.userId, $((c.pageNum-1) * ThreadsPerPage), $ThreadsPerPage):
```
Comment (dom96) [inline forms.tmpl]: 80 chars per line please :P

File: ranks.nim [hunk 6]
```diff
@@ -0,0 +1,12 @@
+
+type
+  Rank* = enum ## serialized as 'status'
+    Spammer          ## spammer: every post is invisible
```
Comment (dom96) [inline ranks.nim]: huh? ``Moderated`` should be "every post is invisible. I don't see the point of this rank.

File: ranks.nim [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+    Troll            ## troll: cannot write new posts
```
Comment (dom96) [inline ranks.nim]: Just call this ``Banned``

File: ranks.nim [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+    Inactive         ## member is not inactive
+    EmailUnconfirmed ## member with unconfirmed email address
```
Comment (dom96) [inline ranks.nim]: Why is there both an ``Inactive`` and an ``EmailUnconfirmed``? They're the same thing.

File: ranks.nim [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+    Moderated        ## new member: posts manually reviewed before everybody
+                     ## can see them
+    User             ## Ordinary user
+    Moderator        ## Moderator: can ban/troll/moderate users
```
Comment (dom96) [inline ranks.nim]: A moderator doesn't "troll" users...

File: forum.nim [hunk 10]
```diff
@@ -699,17 +704,24 @@ proc verifyIdentHash(c: var TForumData, name, epoch, ident: string): bool =
...
 
+proc setStatus(c: var TForumData, nick: string, status: Rank;
+               reason: string): bool =
+  const query =
+    sql("update person set status = ?, ban = ? where name = ?")
+  result = tryExec(db, query, $status, reason, nick)
+  when false:
+    # for now we filter Spammers in forms.tmpl, so that a moderator
+    # cannot accidentically delete all of a user's posts. We go even
+    # further than that and show spammers their own spam postings.
```
Comment (dom96) [inline forum.nim]: yeah... this is a bad idea. Remove it.

File: editdb.nim (hunk 1 above)
Comment (Araq) [inline editdb.nim]: Spammer: no post is visible.
Troll: has been banned, posts visible.

Besides, there is the obvious semantic distinction.

File: forms.tmpl (hunk 3 above)
Comment (Araq) [inline forms.tmpl]: it's impossible to say without benchmarking and it's the logical solution.

File: forms.tmpl (hunk 4 above)
Comment (Araq) [inline forms.tmpl]: No, it's hard enough already to get the logic right.

File: forms.tmpl (hunk 5 above)
Comment (Araq) [inline forms.tmpl]: The other code doesn't adhere to this rule either. The line I changed had the very same length.

File: ranks.nim (hunk 6 above)
Comment (Araq) [inline ranks.nim]: No, there is a semantic difference. 'Moderated' means the user can write new posts. A 'spammer' cannot.

File: ranks.nim (hunk 7 above)
Comment (Araq) [inline ranks.nim]: No, Spammers are also banned, but not trolls.

File: ranks.nim (hunk 8 above)
Comment (Araq) [inline ranks.nim]: Not the same thing, but 'inactive' is not used really, so ok.

File: forum.nim (hunk 10 above)
Comment (Araq) [inline forum.nim]: What the...? Why is this a bad idea?

File: ranks.nim (hunk 7 above)
Comment (dom96) [inline ranks.nim]: this distinction just complicates things, simplify it please.

Comment (Araq) [inline ranks.nim]: delete * from posts where autor is a spammer

is a common operation we want to do to cleanup the database. I cannot do this when we conflate spammers and trolls. Remember: Trolls can have valuable posts before they become trolls.

Comment (dom96) [inline ranks.nim]: This is a UI thing. There is no reason for keeping a spammers posts, only to then later on delete them all after they are marked as a ``Spammer``. Just add a button that does both operations: bans and deletes all posts.

**Pull request merged**